src/analysis/schwartz_space.lean

```diff
@@ -0,0 +1,383 @@
+-/
+
+noncomputable theory
+
+variables {𝕜 𝕜' E F : Type*}
+
+variables [normed_add_comm_group E] [normed_space ℝ E]
+variables [normed_add_comm_group F] [normed_space ℝ F]
+
+variables (E F)
+
+/-- A function is a Schwartz function if it is smooth and all derivatives decay faster than
+  any power of `∥x∥`. -/
+structure schwartz_map :=
+  (to_fun : E → F)
+  (smooth' : cont_diff ℝ ⊤ to_fun)
+  (decay' : ∀ (k n : ℕ), ∃ (C : ℝ), ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n to_fun x∥ ≤ C)
+
+localized "notation `𝓢(` E `, ` F `)` := schwartz_map E F" in schwartz_space
+
+variables {E F}
+
+namespace schwartz_map
+
+instance : has_coe 𝓢(E, F) (E → F) := ⟨to_fun⟩
+
+instance fun_like : fun_like 𝓢(E, F) E (λ _, F) :=
+{ coe := λ f, f.to_fun,
+  coe_injective' := λ f g h, by cases f; cases g; congr' }
+
+/-- Helper instance for when there's too many metavariables to apply `fun_like.has_coe_to_fun`. -/
+instance : has_coe_to_fun 𝓢(E, F) (λ _, E → F) := ⟨λ p, p.to_fun⟩
+
+/-- All derivatives of a Schwartz function are rapidly decaying. -/
+lemma decay (f : 𝓢(E, F)) (k n : ℕ) : ∃ (C : ℝ) (hC : 0 < C),
+  ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ C :=
+begin
+  rcases f.decay' k n with ⟨C, hC⟩,
+  exact ⟨max C 1, by positivity, λ x, (hC x).trans (le_max_left _ _)⟩,
+end
+
+/-- Every Schwartz function is smooth. -/
+lemma smooth (f : 𝓢(E, F)) (n : ℕ∞) : cont_diff ℝ n f := f.smooth'.of_le le_top
+
+@[ext] lemma ext {f g : 𝓢(E, F)} (h : ∀ x, (f : E → F) x = g x) : f = g := fun_like.ext f g h
+
+section aux
+
+lemma bounds_nonempty (k n : ℕ) (f : 𝓢(E, F)) :
+  ∃ (c : ℝ), c ∈ {c : ℝ | 0 ≤ c ∧ ∀ (x : E), ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c} :=
+let ⟨M, hMp, hMb⟩ := f.decay k n in ⟨M, le_of_lt hMp, hMb⟩
+
+lemma bounds_bdd_below (k n : ℕ) (f : 𝓢(E, F)) :
+  bdd_below {c | 0 ≤ c ∧ ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c} :=
+⟨0, λ _ ⟨hn, _⟩, hn⟩
+
+lemma decay_add_le_aux (k n : ℕ) (f g : 𝓢(E, F)) (x : E) :
+  ∥x∥^k * ∥iterated_fderiv ℝ n (f+g) x∥ ≤
+  ∥x∥^k * ∥iterated_fderiv ℝ n f x∥
+  + ∥x∥^k * ∥iterated_fderiv ℝ n g x∥ :=
+begin
+  rw ←mul_add,
+  refine mul_le_mul_of_nonneg_left _ (by positivity),
+  convert norm_add_le _ _,
+  exact iterated_fderiv_add_apply (f.smooth _) (g.smooth _),
+end
+
+lemma decay_neg_aux (k n : ℕ) (f : 𝓢(E, F)) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n (-f) x∥ = ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ :=
+begin
+  nth_rewrite 3 ←norm_neg,
+  congr,
+  exact iterated_fderiv_neg_apply,
+end
+
+variables [normed_field 𝕜] [normed_space 𝕜 F] [smul_comm_class ℝ 𝕜 F]
+
+lemma decay_smul_aux (k n : ℕ) (f : 𝓢(E, F)) (c : 𝕜) (x : E) :
+  ∥x∥ ^ k * ∥iterated_fderiv ℝ n (c • f) x∥ =
+  ∥c∥ * ∥x∥ ^ k * ∥iterated_fderiv ℝ n f x∥ :=
+by rw [mul_comm (∥c∥), mul_assoc, iterated_fderiv_const_smul_apply (f.smooth _), norm_smul]
+
+end aux
+
+section seminorm_aux
+
+/-- Helper definition for the seminorms of the Schwartz space. -/
+@[protected]
+def seminorm_aux (k n : ℕ) (f : 𝓢(E, F)) : ℝ :=
+Inf {c | 0 ≤ c ∧ ∀ x, ∥x∥^k * ∥iterated_fderiv ℝ n f x∥ ≤ c}
```

Comment:
Did you consider defining/characterizing this as the pullback of `norm_seminorm` on `bounded_continuous_function` by the linear map sending `f` to `λ x, ∥x∥^k • iterated_fderiv ℝ n f x`? You may be able to reduce the amount of auxiliary lemmas, but maybe you'd lose it in boilerplate

Reply:
I was basically copying the approach taken by the operator norm and my guess was if it is good enough for the operator norm then it is good enough for the Schwartz space.